compiler/src/iree/compiler/Dialect/Flow/Transforms/FormScalarDispatches.cpp

```diff
@@ -0,0 +1,279 @@
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/TopologicalSortUtils.h"
+
+#define DEBUG_TYPE "iree-flow-form-scalar-dispatches"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+/// Pass declaration.
+struct FormScalarDispatchesPass
+    : public FormScalarDispatchesBase<FormScalarDispatchesPass> {
+  using FormScalarDispatchesBase<
+      FormScalarDispatchesPass>::FormScalarDispatchesBase;
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<affine::AffineDialect, IREE::Flow::FlowDialect,
+                    linalg::LinalgDialect, tensor::TensorDialect>();
+  }
+
+  void runOnOperation() override;
+};
+}  // namespace
+
+/// Return true if type represents a value less than `n` elements.
+static bool isScalarOrTensorOfLinearSizeN(int n, Type type) {
+  if (type.isIntOrIndexOrFloat()) {
+    return true;
+  }
+  if (auto tensorType = type.dyn_cast<RankedTensorType>()) {
+    if (!tensorType.hasStaticShape()) {
+      return false;
+    }
+    return tensorType.getNumElements() <= n;
+  }
+  return false;
+}
+
+/// Return `true` for operations that are to be treated as compute operations.
+static bool isComputeOperation(Operation *op) {
+  MLIRContext *context = op->getContext();
+  if (op->getDialect() == context->getLoadedDialect<linalg::LinalgDialect>()) {
+    return true;
+  }
+  if (op->getDialect() == context->getLoadedDialect<tensor::TensorDialect>()) {
+    return !isa<tensor::ExpandShapeOp, tensor::CollapseShapeOp, tensor::CastOp>(
+        op);
+  }
+  return false;
+}
+
+/// Return `true` if the workload of this operation is less than `n`.
+static bool isOperationWorkloadLessThanSizeN(int n, Operation *candidateOp) {
+  return llvm::all_of(candidateOp->getOperands(),
+                      [&](Value v) {
+                        return isScalarOrTensorOfLinearSizeN(n, v.getType());
+                      }) &&
+         llvm::all_of(candidateOp->getResultTypes(), [&](Type t) {
+           return isScalarOrTensorOfLinearSizeN(n, t);
+         });
+}
+
+/// Return `true` is the operation is to be treated as a scalar operation
+/// and moved into a scalar dispatch (not necessarily as the root of the
+/// dispatch).
+static bool isScalarOperation(int workload, Operation *op) {
+  // 1. Ignore most operations. Only look for a whitelist set of operations.
+  if (!isComputeOperation(op)) {
+    return false;
+  }
+
+  // 2. Check that the workload of the operation is less then the limit
+  if (!isOperationWorkloadLessThanSizeN(workload, op)) {
+    return false;
+  }
+
+  // 3. Do not move operations that are cloned into the dispatch region.
```

Comment:
what's this mean? isScalarOperation(op) is true if it's (**NOT** clonable into dispatch **OR** has one use) feels odd to me? (things clonable into dispatch regions should be put into dispatch regions, and multi-use seems pretty common from what I saw?)

maybe add a test for this specific case showing why it's not expected to work?

Reply:
Basically we want to undo CSE of operations that really should be cloned in each dispatch (`tensor.empty`, `affine.*` ops, even some `linalg.fill` operations). If we just move them into dispatch, they will unnecessarily become outputs of a dispatch and need allocation etc.... So operations that are not really "compute" are clonable (`linalg.fill` is on the margins here...).  This will avoid false SSA dependency between operations.